[fullscreen] Remove hierarchy requirement from fullscreen ready check

nt1m opened the pull request:
#### 675766f7450526b44c5f863da3d42ddd3aa9a13d
<pre>
[fullscreen] Remove hierarchy requirement from fullscreen ready check
<a href="https://bugs.webkit.org/show_bug.cgi?id=257199">https://bugs.webkit.org/show_bug.cgi?id=257199</a>
rdar://110004138

Reviewed by Darin Adler.

The hierarchy requirement was part of the old W3C spec, but is no longer part of the current WHATWG spec. It has also been removed from Blink &amp; Gecko.

Also fix a nullptr deref to make popover-top-layer-interactions.html pass, by adding step 11 of exit fullscreen algorithm to both ExitMode codepaths.

* LayoutTests/fullscreen/full-screen-restrictions-expected.txt:
* LayoutTests/fullscreen/full-screen-restrictions.html:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-ready-check-fullscreen-element-sibling-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-non-top-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-two-elements-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/fullscreen-reordering-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/fullscreen-reordering.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-top-layer-interactions-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/semantics/popovers/popover-top-layer-interactions-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-top-layer-interactions-expected.txt.
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::exitFullscreen): Fix nullptr deref in the edge case where 2 exit fullscreen event loop tasks are queued next to each other.

Canonical link: <a href="https://commits.webkit.org/266426@main">https://commits.webkit.org/266426@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d46c717a2819813ff01b5f6ad8fee71365ee0e47

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13836 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14150 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14484 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15572 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13145 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13917 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16658 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14232 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15819 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14003 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14623 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11724 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16277 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11910 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12484 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19524 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12986 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/12648 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15864 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13179 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11060 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12450 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/12345 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/3359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16782 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1596 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13022 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/fullscreen-reordering.html:
```diff
@@ -0,0 +1,62 @@
+<script src="/resources/testharnessreport.js"></script>
+<script src="/resources/testdriver.js"></script>
+<script src="/resources/testdriver-vendor.js"></script>
+<script src="../../html/semantics/popovers/resources/popover-utils.js"></script>
```
nit: as a followup, it might be good to make this into an ES module... then it's more explicit where isTopLayer and blessTopLayer are coming from.

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/fullscreen-reordering.html:
```diff
@@ -0,0 +1,62 @@
+  assert_true(!isTopLayer(A) && !isTopLayer(B));
+  await blessTopLayer(document.body);
+  await A.requestFullscreen();
+  assert_equals(document.fullscreenElement,A,'first A request');
```
nit, white space between the comas... might be good to tidy this consistently

nt1m replied:
this is a test imported from upstream, I can clean it up upstream later

marcoscaceres reviewed (approved):
non-reviewers approval from me.

nt1m committed:
[fullscreen] Remove hierarchy requirement from fullscreen ready check
https://bugs.webkit.org/show_bug.cgi?id=257199
rdar://110004138

Reviewed by Darin Adler.

The hierarchy requirement was part of the old W3C spec, but is no longer part of the current WHATWG spec. It has also been removed from Blink & Gecko.

Also fix a nullptr deref to make popover-top-layer-interactions.html pass, by adding step 11 of exit fullscreen algorithm to both ExitMode codepaths.

* LayoutTests/fullscreen/full-screen-restrictions-expected.txt:
* LayoutTests/fullscreen/full-screen-restrictions.html:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-ready-check-fullscreen-element-sibling-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-non-top-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-two-elements-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/fullscreen-reordering-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/fullscreen-reordering.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-top-layer-interactions-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/semantics/popovers/popover-top-layer-interactions-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-top-layer-interactions-expected.txt.
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::exitFullscreen): Fix nullptr deref in the edge case where 2 exit fullscreen event loop tasks are queued next to each other.

Canonical link: https://commits.webkit.org/266426@main
